### DISCUSSION TITLE
Can 'Shelve' and 'dbm' not be used in Android?

### DISCUSSION THREAD

**Discussion by jlij** [beeware/toga#2795] [Q&A] [answered]
Does Toga have its own caching API？

**Comment by jlij** [comment]
I tried using JSON, but it was not successful.

**Comment by freakboy3742** [comment]
No, Toga doesn't have a caching API. Toga is a GUI framework; it leaves tasks that aren't related to GUIs, or device hardware related to GUIs (such as GPS), to other libraries.

And yes - you can use shelve, dbm, JSON, SQLite, or any other builtin data storage library with a Beeware/Toga app on Android. If you're having difficulties, the issue is almost certainly the specification of the filesystem location that you're attempting to use to save the file; however, without details, we can't provide any advice on how to fix problems that you may be having.

**Reply by jlij** [reply]
E/AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.testoga/org.beeware.android.MainActivity}: com.chaquo.python.PyException: OSError: [Errno 30] Read-only file system: b'mydata.db.dat'

import toga
import shelve
from toga.style import Pack
from toga.style.pack import COLUMN, ROW


class testoga(toga.App):

    def shelve_test(self):
        with shelve.open('mydata.db') as db:
            # 存储一个字典
            db['user'] = {'name': 'Alice', 'age': 30}

    def startup(self):
        self.shelve_test()
        main_box = toga.Box()
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()


def main():
    return testoga()

**Reply by freakboy3742** [reply]
As I indicated, the issue is filesystem location. You're specifying a file of `mydata.db`. That is a relative file location, and as a result, Python will assume the file is in your current working directory. On Android, your current working directory *is not a writable location*. 

You need to specify a *writeable* file system location, preferably as an absolute path. Since this is a Toga app, the file `self.paths.app.data / "mydata.db"` will give you a writable location, although you may need to call `self.paths.app.data.mkdir(exist_ok=True, parents=True)` first to ensure that the directory has been created.

**Reply by jlij** [reply]
Can you give me some examples?

**Reply by freakboy3742** [reply, marked as answer]
I *have* given you an example. Instead of calling:

    with shelve.open('mydata.db'):
        ...

call:

    self.paths.data.mkdir(exist_ok=True, parents=True)
    with shelve.open(self.paths.data / "mydata.db"):
        ...

**Reply by jlij** [reply]
Please forgive my foolishness, it did not materialize, please help me correct it.



import toga
import shelve
from toga import paths


class testoga(toga.App):

    def test_path(self):
        paths.Paths.app.mkdir(exist_ok=True, parents=True)
        with shelve.open(paths.Paths.app.data / "mydata.db") as db:
            db['key1'] = b'This is a value'

            value = db['key1']
            print(value)

    def startup(self):
        self.test_path()
        main_box = toga.Box()
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()


def main():
    return testoga()

**Reply by freakboy3742** [reply]
I can't fix problems if you don't tell me what error you're seeing. 

However, as a first guess - the code you've presented *isn't* what I told you to write. The example code I gave you uses `self.paths`, not `path.Paths`. I'll admit there's also a typo in the code that I provided; I've corrected that typo.

**Reply by jlij** [reply]
Thank you very much for your patient guidance

**Comment by jlij** [comment]
--------- beginning of crash
E/AndroidRuntime: FATAL EXCEPTION: main
E/AndroidRuntime: Process: com.example.testoga, PID: 10803
E/AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.testoga/org.beeware.android.MainActivity}: com.chaquo.python.PyException: OSError: [Errno 30] Read-only file system: b'mydb.dbm.dat'

import toga
import dbm
from toga.style import Pack
from toga.style.pack import COLUMN, ROW


class testoga(toga.App):

    with dbm.open('mydb.dbm', 'c') as db:
        # 存储数据
        db['key1'] = b'This is a value'

        value = db['key1']
        print(value)

    def startup(self):
        main_box = toga.Box()
        self.main_window = toga.MainWindow(title=self.formal_name)
        self.main_window.content = main_box
        self.main_window.show()


def main():
    return testoga()
